### ISSUE TITLE
OutOfMemory error when recovering with a large number of snapshots

### ISSUE THREAD

**Issue by cridmann** [dnvriend/akka-persistence-jdbc#17]
Hello,

I was doing some performance testing, and 1 one of my persistent actors makes heavy use of snapshotting.  I'm getting an OutOfMemory error when I try to start up this persistent actor (at app startup time) and it tries to recover.  I am using Oracle, Akka 2.3.9, and akka-persistence-jdbc 1.1.1

I believe the problem code is:

```
  def selectSnapshotsFor(persistenceId: String, criteria: SnapshotSelectionCriteria): List[SelectedSnapshot] =
    SQL(s"SELECT * FROM $schema$table WHERE persistence_id = ? AND sequence_nr <= ? ORDER BY sequence_nr DESC")
      .bind(persistenceId, criteria.maxSequenceNr)
      .map { rs => SelectedSnapshot(SnapshotMetadata(rs.string("persistence_id"), rs.long("sequence_nr"), rs.long("created")), unmarshal(rs.string("snapshot")).data) }
      .list()
      .apply()
      .filterNot(snap => snap.metadata.timestamp > criteria.maxTimestamp)
```

Here is the JDBC call it tries to make:
```
SELECT * FROM SNAPSHOT WHERE persistence_id = 'myPersistenceId' AND sequence_nr <= 9223372036854775807 ORDER BY sequence_nr DESC;
```

It appears that this code is reading all of the snapshots into memory and filtering by timestamp only after.  Why not put this filter within the query itself?

Here is the stack-trace:

```
[ERROR] - from akka.actor.ActorSystemImpl in application-akka.actor.default-dispatcher-4
Uncaught error from thread [application-akka.actor.default-dispatcher-16] shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled
java.lang.OutOfMemoryError: Java heap space
        at java.lang.reflect.Array.newArray(Native Method) ~[na:1.8.0_11]
        at java.lang.reflect.Array.newInstance(Array.java:75) ~[na:1.8.0_11]
        at oracle.jdbc.driver.BufferCache.get(BufferCache.java:226) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at oracle.jdbc.driver.PhysicalConnection.getByteBuffer(PhysicalConnection.java:7659) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at oracle.jdbc.driver.T4C8TTIClob.read(T4C8TTIClob.java:226) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at oracle.jdbc.driver.T4CConnection.getChars(T4CConnection.java:3187) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at oracle.sql.CLOB.getChars(CLOB.java:459) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at oracle.sql.CLOB.getSubString(CLOB.java:321) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at oracle.jdbc.driver.T4CClobAccessor.getString(T4CClobAccessor.java:474) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at oracle.jdbc.driver.OracleResultSetImpl.getString(OracleResultSetImpl.java:1297) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at oracle.jdbc.driver.OracleResultSet.getString(OracleResultSet.java:494) ~[com.oracle.ojdbc6-11.2.0.3.0.jar:11.2.0.3.0]
        at scalikejdbc.TypeBinder$$anonfun$57.apply(TypeBinder.scala:112) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.TypeBinder$$anonfun$57.apply(TypeBinder.scala:112) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.TypeBinder$$anon$2.apply(TypeBinder.scala:43) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.WrappedResultSet$$anonfun$get$2.apply(WrappedResultSet.scala:474) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.WrappedResultSet.wrapIfError(WrappedResultSet.scala:40) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.WrappedResultSet.get(WrappedResultSet.scala:474) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.WrappedResultSet.string(WrappedResultSet.scala:361) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at akka.persistence.jdbc.snapshot.GenericStatements$$anonfun$1.apply(Statements.scala:50) ~[com.github.dnvriend.akka-persistence-jdbc_2.11-1.1.1.jar:1.1.1]
        at akka.persistence.jdbc.snapshot.GenericStatements$$anonfun$1.apply(Statements.scala:48) ~[com.github.dnvriend.akka-persistence-jdbc_2.11-1.1.1.jar:1.1.1]
        at scalikejdbc.DBSession$$anonfun$traversable$1$$anonfun$apply$2.apply(DBSession.scala:311) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.DBSession$$anonfun$traversable$1$$anonfun$apply$2.apply(DBSession.scala:311) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245) ~[org.scala-lang.scala-library-2.11.6.jar:na]
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245) ~[org.scala-lang.scala-library-2.11.6.jar:na]
        at scalikejdbc.ResultSetTraversable$$anonfun$foreach$1.apply(ResultSetTraversable.scala:37) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.ResultSetTraversable$$anonfun$foreach$1.apply(ResultSetTraversable.scala:34) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.LoanPattern$class.using(LoanPattern.scala:33) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.ResultSetTraversable.using(ResultSetTraversable.scala:24) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.ResultSetTraversable.foreach(ResultSetTraversable.scala:34) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:245) ~[org.scala-lang.scala-library-2.11.6.jar:na]
        at scalikejdbc.ResultSetTraversable.map(ResultSetTraversable.scala:24) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
        at scalikejdbc.DBSession$$anonfun$traversable$1.apply(DBSession.scala:311) ~[org.scalikejdbc.scalikejdbc-core_2.11-2.2.4.jar:2.2.4]
```

**Comment by dnvriend** [comment]
I released v1.1.3 that should do a better job handing snapshots and selecting one. I profiled the plugin and it should now do a better job. Could you please try whether or not you still have the out of memory problem?

**Comment by cridmann** [comment]
Thanks for the quick turn-around!  I'll test it in the next few days once it gets in the right environment - stay tuned.
